api/logs/v1/rules_labels_enforcer.go

```diff
@@ -0,0 +1,410 @@
+	Rules    []rule  `json:"rules"`
+	Interval float64 `json:"interval"`
+}
+
+type rule struct {
+	*alertingRule
+	*recordingRule
+}
+
+func (r *rule) Labels() labels.Labels {
+	if r.alertingRule != nil {
+		return r.alertingRule.Labels
+	}
+	return r.recordingRule.Labels
+}
+
+// MarshalJSON implements the json.Marshaler interface for rule.
+func (r *rule) MarshalJSON() ([]byte, error) {
+	if r.alertingRule != nil {
+		return json.Marshal(r.alertingRule)
+	}
+	return json.Marshal(r.recordingRule)
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface for rule.
+func (r *rule) UnmarshalJSON(b []byte) error {
+	var ruleType struct {
+		Type string `json:"type"`
+	}
+	if err := json.Unmarshal(b, &ruleType); err != nil {
+		return err
+	}
+	switch ruleType.Type {
+	case "alerting":
+		var alertingr alertingRule
+		if err := json.Unmarshal(b, &alertingr); err != nil {
+			return err
+		}
+		r.alertingRule = &alertingr
+	case "recording":
+		var recordingr recordingRule
+		if err := json.Unmarshal(b, &recordingr); err != nil {
+			return err
+		}
+		r.recordingRule = &recordingr
+	default:
+		return fmt.Errorf("failed to unmarshal rule: unknown type %q", ruleType.Type)
+	}
+
+	return nil
+}
+
+type rulesData struct {
+	RuleGroups []*ruleGroup `json:"groups,omitempty"`
+	Alerts     []*alert     `json:"alerts,omitempty"`
+}
+
+type prometheusRulesResponse struct {
+	Status    string    `json:"status"`
+	Data      rulesData `json:"data"`
+	Error     string    `json:"error"`
+	ErrorType string    `json:"errorType"`
+}
+
+type lokiRule struct {
+	Alert       string            `json:"alert,omitempty"`
+	Record      string            `json:"record,omitempty"`
+	Expr        string            `json:"expr,omitempty"`
+	For         string            `json:"for,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Labels      map[string]string `json:"labels,omitempty"`
+}
+
+type lokiRuleGroup struct {
+	Name     string     `json:"name"`
+	Interval string     `json:"interval,omitempty"`
+	Limit    int        `json:"limit,omitempty"`
+	Rules    []lokiRule `json:"rules"`
+}
+
+type lokiRulesResponse = map[string][]lokiRuleGroup
+
+func WithEnforceRulesLabelFilters(labelKeys map[string][]string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			tenant, ok := authentication.GetTenant(r.Context())
+			if !ok {
+				httperr.PrometheusAPIError(w, "missing tenant id", http.StatusBadRequest)
+
+				return
+			}
+
+			keys, ok := labelKeys[tenant]
+			if !ok || len(keys) == 0 {
+				next.ServeHTTP(w, r)
+
+				return
+			}
+
+			// Check that the user provides each label filter as a non-empty URL parameter
+			queryParams := r.URL.Query()
+			for _, key := range keys {
+				if !queryParams.Has(key) {
+					httperr.PrometheusAPIError(w, fmt.Sprintf("missing URL parameter %s", key), http.StatusBadRequest)
```

Comment:
If I understand correctly, this makes it mandatory to use all the defined labels. I don't think this is a good idea. 
I think we should be able to define a set of labels to filter on, and if some or none of those labels are provided, then we should still be able to query.

Reply:
I like to think this the other way around. They `keys` are provided by the administrator via `--logs.rules.label-filters` to ensure that users only search for logs giving mandatory input to these `keys` so that our middleware can enforce authorization constraints for the given input.

Reply:
Thing is, they are not really provided by the administrator. This something the Operator defines in our case. 

In terms of API usage, if we set for instance two keys A and B, we no longer have the ability to use the API without these two filters, and we are also forced to use both A and B filters in order to get a response.
This in my opinion makes the API too restrictive, because even as a cluster admin I cannot query without all of these per-defined filters.

Reply:
As agreed on our last call, I am writing the following down for the record. The behavior we actually want is twofold:
- For users with authorization matchers returned from the authz endpoint, we should return an empty list if the user is not providing any rule label filters per URL query parameters.
- For users without authorization matchers, we should return the original response if the user is not providing any rule label filters per URL query parameters.